lib/plug/conn.ex

```diff
@@ -544,7 +544,9 @@ defmodule Plug.Conn do
 
   def put_req_header(%Conn{adapter: adapter, req_headers: headers} = conn, key, value) when
       is_binary(key) and is_binary(value) do
-    validate_header_key!(adapter, key)
+    if Application.get_env(:plug, :validate_header_keys_during_test, true) do
```

Comment:
My suggestion is to rename the function to `validate_header_key_if_test!` and move the check there.